docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,348 @@
+
+### Opportunities for Simplification
+
+**Claim:** There are too many IPC mechanisms.
+
+The preponderance of ABCI usage is via the Cosmos SDK, which means the
+application and the consensus node are compiled together into a single binary,
+and the consensus node calls the ABCI methods of the application directly as Go
+functions.
+
+TODO: What are the remaining use cases for the socket protocol and gRPC? Ethan
+mentioned some cases like web-based explorer UIs and possibly other tools.
+
+As far as I can tell the primary consumers of the multi-headed "RPC service"
+are the light client and the `tendermint` command-line client. There is
+probably some local use via curl, but I expect that is mostly ad hoc. Ethan
+reports that nodes are often configured with the ports to the RPC service
+blocked, which is good for security but complicates use by the light client.
+
+TODO: Figure out what is going on with the weird pseudo-inheritance service
+interface.  The BaseService type makes sense, but its embeds are more
+complicated.
+
+### Context: Remote Signer Issues
+
+Since the remote signer needs a secure communication channel to exchange keys
+and signatures, and is expected to run truly remotely from the node (i.e., on a
+separate physical server), there is not a whole lot we can do here. We should
+finish the deprecation and removal of the "raw" socket protocol between the
+consensus node and remote signers, but the use of gRPC is appropriate.
+
+The main improvement we can make is to simplify the implementation quite a bit,
+once we no longer need to support both "raw" and gRPC transports.
+
+### Context: ABCI Issues
+
+In the original design of ABCI, the presumption was that all access to the
+application should be mediated by the consensus node. The idea is that outside
+access could change application state and corrupt the consensus process, which
+relies on the application to be deterministic. Of course, even without outside
+access an application could behave nondeterministically, but allowing other
+programs to send it requests was seen as courting trouble.
+
+Conversely, users noted that most of the time, tools written for a particular
+application don't want to talk to the consensus module directly. The
+application "owns" the state machine the consensus engine is replicating, so
+tools that care about application state should talk to the application.
+Otherwise, they would have to bake in knowledge about Tendermint (e.g., its
+interfaces and data structures) just because of the mediation.
+
+For clients to talk directly to the application, however, there is another
+concern: The consensus node is the ABCI _client_, so it is inconvenient for the
+application to "push" work into the consensus module via ABCI itself.  The
+current implementation works around this by calling the consensus node's RPC
+service, which exposes an `ABCIQuery` kitchen-sink method that allows the
+application a way to poke ABCI messages in the other direction.
```

Comment:
My personal thoughts are to remove ABCI queries. The cosmos SDK which by proxy is our main user already comes with ways to query the application state. Even most other implementation have their own gRPC server or similar. Having multiple ways to do things is confusing (and very much the crux of this RFC).

Reply:
Yes, I'd like to be able to do that. I think Cosmos SDK already exposes a gRPC server for other tools to talk to the application directly. But I _think_ the SDK may still rely on this `ABCIQuery` endpoint to implement adding a transaction received from an outside gRPC client (unless I misread the code). So we'll have some work to do first.

Reply:
@marbar3778 I seem to recall in the blur of issues I scroll through that the CosmosSDK were planning on removing dependency on `ABCIQuery`. Can you confirm this or is it just a figment of my imagination.